lisa/sut_orchestrator/azure/tools.py

```diff
@@ -586,3 +514,117 @@ def get_pool_records(self, pool_id: int, force_run: bool = False) -> Dict[str, s
             records[content_split[i]] = content_split[i + 1]
 
         return records
+
+
+class AzureBlobOperator(Tool):
+    @property
+    def command(self) -> str:
+        return "echo azure_blob_operator"
+
+    @property
+    def can_install(self) -> bool:
+        return False
+
+    def _check_exists(self) -> bool:
+        return True
+
+    def download_blob(
+        self,
+        account_name: str,
+        container_name: str,
+        blob_name: str,
+        cloud: Cloud = AZURE_PUBLIC_CLOUD,
+        credential: Optional[Any] = None,
+        subscription_id: Optional[str] = None,
+        resource_group_name: Optional[str] = None,
+        connection_string: Optional[str] = None,
+    ) -> PurePath:
+        working_path = constants.RUN_LOCAL_WORKING_PATH
+        working_path.mkdir(parents=True, exist_ok=True)
+        file_path = working_path / blob_name
+        container_client = get_or_create_storage_container(
+            container_name,
+            credential,
+            subscription_id,
+            cloud,
+            account_name,
+            resource_group_name,
+            connection_string,
+            False,
+        )
+        blob_client = container_client.get_blob_client(blob_name)
+        if not blob_client.exists():
+            raise LisaException(
+                f"Blob {blob_name} not found in container {container_name}"
+            )
+        try:
+            with global_download_blob_lock:
+                blob_properties = blob_client.get_blob_properties()
+                if file_path.exists():
+                    blob_hash = blob_properties.content_settings.content_md5
+                    if blob_hash:
+                        blob_hash_hex = "".join(f"{byte:02x}" for byte in blob_hash)
+                        if self._calculate_hash(file_path) == blob_hash_hex:
+                            self._log.debug(
+                                f"Blob {blob_name} already exists in {file_path}. "
+                                "No need to download again."
+                            )
+                            return file_path
+
+                blob_size = blob_properties["size"]
+                downloaded_size = 0
+                with open(file_path, "wb") as file:
+                    download_stream = blob_client.download_blob()
+                    for chunk in download_stream.chunks():
+                        file.write(chunk)
+                        downloaded_size += len(chunk)
+                        percentage = (downloaded_size / blob_size) * 100
+                        self._log.debug(
```

Comment:
Make sure it won't print too many lines. For example, it's ok to update every 10 seconds.

Reply:
updated